presto-main/src/main/java/com/facebook/presto/server/InternalAuthenticationManager.java

```diff
@@ -0,0 +1,142 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server;
+
+import com.facebook.airlift.http.client.HttpRequestFilter;
+import com.facebook.airlift.http.client.Request;
+import com.facebook.airlift.http.server.BasicPrincipal;
+import com.facebook.airlift.log.Logger;
+import com.facebook.airlift.node.NodeInfo;
+import com.facebook.presto.server.security.SecurityConfig;
+import com.google.common.hash.Hashing;
+import io.jsonwebtoken.JwtException;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+
+import java.security.Principal;
+import java.time.ZonedDateTime;
+import java.util.Date;
+import java.util.Optional;
+
+import static com.facebook.airlift.http.client.Request.Builder.fromRequest;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class InternalAuthenticationManager
+        implements HttpRequestFilter
+{
+    private static final Logger log = Logger.get(InternalAuthenticationManager.class);
+
+    private static final String PRESTO_INTERNAL_BEARER = "X-Presto-Internal-Bearer";
+
+    private final boolean internalJwtEnabled;
+    private final byte[] hmac;
+    private final String nodeId;
+
+    @Inject
+    public InternalAuthenticationManager(InternalCommunicationConfig internalCommunicationConfig, SecurityConfig securityConfig, NodeInfo nodeInfo)
+    {
+        this(getSharedSecret(internalCommunicationConfig, nodeInfo, !securityConfig.getAuthenticationTypes().isEmpty()), nodeInfo.getNodeId());
+    }
+
+    public InternalAuthenticationManager(Optional<String> sharedSecret, String nodeId)
+    {
+        requireNonNull(sharedSecret, "sharedSecret is null");
+        requireNonNull(nodeId, "nodeId is null");
+        this.internalJwtEnabled = sharedSecret.isPresent();
+        if (internalJwtEnabled) {
+            this.hmac = Hashing.sha256().hashString(sharedSecret.get(), UTF_8).asBytes();
+        }
+        else {
+            this.hmac = null;
+        }
+        this.nodeId = nodeId;
+    }
+
+    private static Optional<String> getSharedSecret(InternalCommunicationConfig internalCommunicationConfig, NodeInfo nodeInfo, boolean authenticationEnabled)
+    {
+        requireNonNull(nodeInfo, "nodeInfo is null");
+
+        // This check should not be required (as bean validation already checked it),
+        // but be extra careful to not use a known secret for authentication.
+        if (!internalCommunicationConfig.isRequiredSharedSecretSet()) {
+            throw new IllegalArgumentException("Shared secret (internal-communication.shared-secret) is required when internal communications uses https");
```

Comment:
isRequiredSharedSecretSet() contains `!isHttpsRequired()` so this exception message doesn't match with that. Do we need to set this for http or https?